keras_cv/models/backbone.py

```diff
@@ -0,0 +1,197 @@
+
+        if load_weights and "weights_url" not in metadata:
+            raise ValueError(
+                f"""Pretrained weights not available for preset "{preset}". """
+                "Set `load_weights=False` to use this preset or choose one of "
+                "the following presets with weights:"
+                f""" "{'", "'.join(cls.presets_with_weights)}"."""
+            )
+
+        config = metadata["config"]
+        model = cls.from_config({**config, **kwargs})
+
+        if not load_weights:
+            return model
+
+        weights = keras.utils.get_file(
+            "model.h5",
+            metadata["weights_url"],
+            cache_subdir=os.path.join("models", preset),
+            file_hash=metadata["weights_hash"],
+        )
+
+        model.load_weights(weights)
+        return model
+
+    def __init_subclass__(cls, **kwargs):
+        # Use __init_subclass__ to setup a correct docstring for from_preset.
+        super().__init_subclass__(**kwargs)
+
+        # If the subclass does not define from_preset, assign a wrapper so that
+        # each class can have an distinct docstring.
+        if "from_preset" not in cls.__dict__:
+
+            def from_preset(calling_cls, *args, **kwargs):
+                return super(cls, calling_cls).from_preset(*args, **kwargs)
+
+            cls.from_preset = classmethod(from_preset)
+
+        # Format and assign the docstring unless the subclass has overridden it.
+        if cls.from_preset.__doc__ is None:
+            cls.from_preset.__func__.__doc__ = Backbone.from_preset.__doc__
+            format_docstring(
+                model_name=cls.__name__,
+                example_preset_name=next(iter(cls.presets_with_weights), ""),
+                preset_names='", "'.join(cls.presets),
+            )(cls.from_preset.__func__)
+
+    @property
+    def backbone_level_outputs(self):
+        """Backbone outputs at each resolution level for transfer learning."""
+        return None
+
+    @backbone_level_outputs.setter
+    def backbone_level_outputs(self, value):
+        self._backbone_level_outputs = value
+
+    def extract_features(self, min_level=None, max_level=None):
+        """Convert the application model into a model backbone for other tasks.
+
+        The backbone model will usually take same inputs as the original
+        application model, but produce multiple outputs, one for each feature
+        level. Those outputs can be feed to network downstream, like FPN and RPN.
+
+        The output of the backbone model will be a dict with int as key and
+        tensor as value. The int key represent the level of the feature output.
+        A typical feature pyramid has five levels corresponding to scales P3,
+        P4, P5, P6, P7 in the backbone. Scale Pn represents a feature map 2n
+        times smaller in width and height than the input image.
+
+        Args:
+            min_level: optional int, the lowest level of feature to be included
```

Comment:
This min_level / max_level API is very mysterious. Doubt anyone not already familiar with the implementation will figure out what it means. Can we find better argument names and descriptions?

Reply:
Good point! Will file an Issue.

Reply:
yeah I really do not like this API.  realistically it should be a list - either of level names or layer names.

Reply:
Filed #1447